### DISCUSSION TITLE
Able to Change Cursor Type for Insert Mode

### DISCUSSION THREAD

**Discussion by lihe07** [helix-editor/helix#9158] [Ideas] [answered]
Most modern terminals support different cursor types. Vim and NeoVim are able to be configured to use different cursor shape in insert mode. (Refer to [this](https://stackoverflow.com/questions/71333744/neovim-change-cursor-type-for-insert-mode) )

I think it is a good feature to distinguish different modes also to indicate the position to insert new characters. A demo of NeoVim is shown below:

<table>
  <tr>
    <td valign="top"><img src="https://github.com/helix-editor/helix/assets/53819558/26fb701e-559b-4973-aa80-adc37664174d"/></td>
    <td valign="top"><img src="https://github.com/helix-editor/helix/assets/53819558/d07e1c7b-6f5e-4246-8fd1-5c3b7323be15"/></td>
  </tr>
  <tr>
   <td>Normal Mode</td>
   <td>Insert Mode</td>
 </tr>
</table>

**Comment by mahalel** [comment, marked as answer]
You can do this by adding the following to your configuration file:

```
[editor.cursor-shape]
insert = "bar"
normal = "block"
select = "underline"
```

See available choices here - https://docs.helix-editor.com/configuration.html#editorcursor-shape-section

**Reply by robinmoussu** [reply]
Is there a way to also modify the cursor for the non-primary cursor in insert mode? ie. `CCCi` will display 3 cursors, the first two are blocks and the third an underline. I’d like all cursors to be underline.

**Reply by the-mikedavis** [reply]
Only the primary cursor can be styled. It's a terminal limitation: bar and underline cursors are styled by using the terminal's cursor and you can only have one terminal cursor

**Reply by robinmoussu** [reply]
Arf. I guess that the other cursors just use a different background color, and that `underline` is not a valid terminal decoration (unlike bold or changing background color)?

**Reply by the-mikedavis** [reply]
Yeah non-primary cursors are drawn with a different background to simulate a block. You could change that highlight to show an underline though, something like:

```toml
"ui.cursor" = { modifiers = ["underlined"] }
```

in your theme file

**Comment by TornaxO7** [comment]
@lihe07 please don't forget to mark your question as answered, if it's solved.
